src/agl/WebAppManagerServiceAGL.cpp

```diff
@@ -714,14 +715,19 @@ void WebAppManagerServiceAGL::onActivateEvent() {
 void
 WebAppManagerServiceAGL::onSendAglEvent()
 {
-	LOG_DEBUG("onSendAglEvent for app_id_event_target_ %s", app_id_event_target_.c_str());
 	if (app_id_event_target_.empty())
 		return;
 
 	WebAppBase* web_app = WebAppManager::instance()->findAppById(app_id_event_target_);
 	if (web_app) {
-		LOG_DEBUG("got webapp for app_id_event_target_ %s, sending ready", app_id_event_target_.c_str());
-		web_app->sendAglReady();
+		WebPageBase *page = web_app->page();
+		LOG_DEBUG("onSendAglEvent for app_id_event_target_ %s", app_id_event_target_.c_str());
+		/* FIXME: there's no assurance that by this point the webapp
+		 * has been up-and-running.  Waiting here would effectively
+		 * block WAM so we need another way of being notified when the
+		 * webapps has been fully loaded */
+		if (page)
```

Comment:
The intended role for sendAglReady is to delay presentation until the client has created its surfaces/loaded its assets etc and is ready to display. The client could create multiple surfaces (2 - 3 panels + background), and we should wait until all of them have been created. 

This translates, ideally, to the fact page/app is fully loaded, the idea being to map a surface to a webapp/webpage. We need to wait for all them and then send the ready() request. 

> Ideally we'd want to notify AGL when this frame is ready, right?

Well, yeah, but we need to wait for all the panels/background.

Reply:
@jdapena could you please point to where exactly is the splash screen  code in WAM/webOS you've mentioned above? As far I can tell, there's no bi-directional communication between the renderer process and the shared process, so I'm not sure how that actually works, unless you already implied that we'd have something equivalent to that such it is eventually delivered to WAM.

Reply:
There are two stages. First, when we get DidSwapCompositorFrame (that means we have all the means to render a frame), we end up calling onStageActivated, that will call Show() on window. This Show() is not expected to swap immediately, but it will activate the chromium GPU frame swaping, so first frame for wayland will come here.

The other important point is that DidSwapCompositorFrame comes from the event DidCompleteSwap... That will be filtered with the code under kUseVizFMPWithTimeout. This will delay first swap to first meaningful paint.